chainio/dispatcher.go

```diff
@@ -0,0 +1,106 @@
+package chainio
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// DefaultProcessBlockTimeout is the timeout value used when waiting for one
+// consumer to finish processing the new block epoch.
+var DefaultProcessBlockTimeout = 60 * time.Second
+
+// ErrProcessBlockTimeout is the error returned when a consumer takes too long
+// to process the block.
+var ErrProcessBlockTimeout = errors.New("process block timeout")
+
+// DispatchSequential takes a list of consumers and notify them about the new
+// epoch sequentially.
+func DispatchSequential(beat Blockbeat, consumers []Consumer) error {
+	for _, c := range consumers {
+		b := beat.copy()
+
+		// Send the copy of the beat to the consumer.
+		err := notifyAndWait(b, c, DefaultProcessBlockTimeout)
+		if err != nil {
+			b.logger().Errorf("Failed to process block: %v", err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DispatchConcurrent notifies each consumer concurrently about the blockbeat.
+func DispatchConcurrent(beat Blockbeat, consumers []Consumer) error {
+	// errChans is a map of channels that will be used to receive errors
+	// returned from notifying the consumers.
+	errChans := make(map[string]chan error, len(consumers))
+
+	// Notify each queue in goroutines.
+	for _, c := range consumers {
+		// Create a signal chan.
+		errChan := make(chan error, 1)
+		errChans[c.Name()] = errChan
+
+		// Notify each consumer concurrently.
+		go func(c Consumer, beat Blockbeat) {
+			b := beat.copy()
+
+			// Send the copy of the beat to the consumer.
+			errChan <- notifyAndWait(
+				b, c, DefaultProcessBlockTimeout,
+			)
+		}(c, beat)
+	}
+
+	// Wait for all consumers in each queue to finish.
+	for name, errChan := range errChans {
```

Comment:
not sure how that helps us fail faster?

Reply:
I don't think we need to have context here - or what are the benefits of using it since we already have a quit chan?

> because as soon as one fails you cancel all the rest and return.

Since we are dispatching them concurrently, each notification should be independent.